igvc_perception/src/multiclass_segmentation/train.py

```diff
@@ -0,0 +1,207 @@
+encoder_weight_decay = 0.00003
+optimizer_weight_decay = 0.0003
+optim_factor = 0.25
+optim_patience = 2
+
+optimizer = AdamW(
+    model.parameters(),
+    lr=0.001,
+    betas=(0.9, 0.999),
+    eps=1e-08,
+    weight_decay=0.01,
+    amsgrad=False,
+)
+
+scheduler = optim.lr_scheduler.ReduceLROnPlateau(
+    optimizer, factor=optim_factor, patience=optim_patience
+)
+
+num_epochs = 10
+device = utils.get_device()
+
+runner = SupervisedRunner(device=device, input_key="image", input_target_key="mask")
+
+# Use Catalyst callbacks for metric calculations during training
+callbacks = [
+    CriterionCallback(input_key="mask", prefix="loss", criterion_key="CE"),
+    MulticlassDiceMetricCallback(input_key="mask"),
+]
+
+# Train and print model training logs
+runner.train(
+    model=model,
+    criterion=criterion,
+    optimizer=optimizer,
+    scheduler=scheduler,
+    loaders=loaders,
+    callbacks=callbacks,
+    logdir="content/full_model2",
+    num_epochs=num_epochs,
+    main_metric="loss",
+    minimize_metric=True,
+    fp16=None,
+    verbose=True,
+)
+
+# Test model on test dataset
+test_data = SegmentationDataset(test_images_path, test_masks_path)
+
+infer_loader = DataLoader(test_data, batch_size=12, shuffle=False, num_workers=4)
+# Get model predictions on test dataset
+predictions = np.vstack(
+    list(
+        map(
+            lambda x: x["logits"].cpu().numpy(),
+            runner.predict_loader(
+                loader=infer_loader, resume=f"content/full_model2/checkpoints/best.pth"
+            ),
+        )
+    )
+)
+
+# Pick sample images to analyze results
+low = 1
+high = len(predictions) - 1
+num_results = 30
+num_rand_results = num_results - 2
+rand_nums = np.random.randint(low, high, num_rand_results)
+
+"""
+The results specifically include images 30 and 141 as
+they are images containing multiple lines and barrels.
+Hence, they would be strong indicators of performance.
+"""
+rand_nums = np.insert(rand_nums, 0, 30)
+rand_nums = np.insert(rand_nums, 0, 141)
+
+# Save the image, mask, and predicted mask for sample test data
+
+for num in rand_nums:
+    # Show and save image
+    image = np.asarray(test_data[num]["image"])
+    image = np.swapaxes(image, 2, 0)
+    image = np.swapaxes(image, 1, 0)
+    image = image.astype(np.uint8)
+    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+    image_label = "Raw_Image_{}".format(num)
+    plt.imsave(image_label + ".png", image)
+
+    # Show and save mask
+    mask = np.squeeze(test_data[num]["mask"])
+    mask_label = "Ground_Truth_{}".format(num)
+    plt.imsave(mask_label + ".png", mask)
+
+    # Show and save predicted mask
+    pred_mask = np.asarray(predictions[num])
+    pred_mask = np.swapaxes(pred_mask, 2, 0)
+    pred_mask = np.swapaxes(pred_mask, 1, 0)
+    pred_mask = pred_mask.astype(np.float64)
+    pred_mask = np.argmax(pred_mask, axis=2)
+    pred_mask_label = "Predicted_Mask_{}".format(num)
+    plt.imsave(pred_mask_label + ".png", pred_mask)
```

Comment:
I'm not sure I like this either.... this results in like a bunch of `Raw_Image_X.png`, `Ground_Truth_X.png` and `Predicted_Mask_X.png` images that are just littered around, and its hard to compare and view.... Maybe something have something like this?

```
predictions_path = "./predictions"
Path(predictions_path).mkdir(parents=True, exist_ok=True)
for num in tqdm(rand_nums):
    fig = plt.figure(figsize=(10, 4))
    # Show and save image
    plt.subplot(1, 3, 1)
    image = np.asarray(test_data[num]["image"])
    image = np.swapaxes(image, 2, 0)
    image = np.swapaxes(image, 1, 0)
    image = image.astype(np.uint8)
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
    plt.imshow(image)
    plt.gca().set_title("Raw Image")

    # Show and save mask
    plt.subplot(1, 3, 2)
    mask = np.squeeze(test_data[num]["mask"])
    plt.imshow(mask)
    plt.gca().set_title("Ground Truth")

    # Show and save predicted mask
    plt.subplot(1, 3, 3)
    pred_mask = np.asarray(predictions[num])
    pred_mask = np.swapaxes(pred_mask, 2, 0)
    pred_mask = np.swapaxes(pred_mask, 1, 0)
    pred_mask = np.argmax(pred_mask, axis=2)
    plt.imshow(pred_mask)
    plt.gca().set_title("Predicted")

    plt.tight_layout()
    plt.savefig(f"./{predictions_path}/{num}.png")
    plt.close()
```

Reply:
Placed it into and ran train.py with the appropriate import

Reply:
I kept the Colab notebook inside the repo as is since it does the triple column and the images are already the output of a cell